### PR TITLE
DSV4 parameter mapping refactor

### PR DESCRIPTION
**Pull request by snehalv2002**
# Description

Currently the DeepSeek V4 parameter mapping maps from the huggingface transformers model structure to HF safetensors checkpoint structure to maxtext structure. This PR changes the mapping to map directly from safetensors checkpoint structure to maxtext structure. The current 2 step mapping is potentially brittle if there are changes to the transformers library and also adds unnecessary complexity.

# Tests

Used `to_maxtext` to convert scanned and unscanned checkpoints. Both passed forward logit test with KL divergence 0.42 and 0.37 respectively. Running training with scanned checkpoint results in `lm_loss` around 2.3. Converted the same checkpoints back to huggingface using `to_huggingface` and the weights matched exactly with the original safetensors checkpoint.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/checkpoint_conversion/to_huggingface.py [hunk 1]
```diff
@@ -279,7 +287,11 @@ def _validate_or_update_architecture(hf_config, max_config, override: bool):
         mt_value = mt_value * 2
 
     # Special handling for Qwen3-MoE: hf.intermediate_size is the aggregated dense MLP dim, but mt.mlp_dim is dim per expert
-    if "qwen3" in max_config.model_name and getattr(max_config, "num_experts", 0) > 1 and hf_attr == "intermediate_size":
+    if (
+        "qwen3" in max_config.model_name
+        and getattr(max_config, "num_experts", 0) > 1
+        and hf_attr == "intermediate_size"
+    ):
```
Comment (snehalv2002) [inline src/maxtext/checkpoint_conversion/to_huggingface.py]: this change is irrelevant to this PR

File: src/maxtext/checkpoint_conversion/to_huggingface.py [hunk 2]
```diff
@@ -430,7 +442,9 @@ def _transform_and_save_weights(
   if lora_restore_path and not load_parameters_path:
     # Adapter Mode
     transformed_hf_weights, found_hf_modules = _transform_weights_to_adapter(param_map, maxtext_state_dict)
-    save_adapter_files(output_directory, transformed_hf_weights, config, found_hf_modules, HF_IDS.get(config.model_name))
+    save_adapter_files(
+        output_directory, transformed_hf_weights, config, found_hf_modules, HF_IDS.get(config.model_name)
+    )
```
Comment (snehalv2002) [inline src/maxtext/checkpoint_conversion/to_huggingface.py]: this change irrelevant to this PR

File: src/maxtext/configs/models/deepseek4-284b.yml [hunk 3]
```diff
@@ -22,6 +22,7 @@ base_mlp_dim: 2048
 base_moe_mlp_dim: 2048
 vocab_size: 129280
 head_dim: 512
+qk_nope_head_dim: 512
```
Comment (snehalv2002) [inline src/maxtext/configs/models/deepseek4-284b.yml]: nope is not used for deepseek. double check whether you used it anywhere, otherwise remove

File: src/maxtext/configs/models/deepseek4-284b.yml [hunk 4]
```diff
@@ -59,6 +60,7 @@ routed_scaling_factor: 1.5
 
 
 # --- Attention configuration ---
+attention: "dot_product"
```
Comment (snehalv2002) [inline src/maxtext/configs/models/deepseek4-284b.yml]: remove this, we recently added flash attention support

File: tests/unit/deepseek_v4_vs_reference_test.py
Comment (snehalv2002) [inline tests/unit/deepseek_v4_vs_reference_test.py:1]: remove any stylistic diffs that we made to this file.

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py [hunk 5]
```diff
@@ -4348,8 +4302,7 @@ def mhc_concat_scale(input_tensors, target_shape=None):
     "deepseek2-16b": DEEPSEEK_MAXTEXT_TO_HF_PARAM_HOOK_FN,
     "deepseek3-671b": DEEPSEEK_MAXTEXT_TO_HF_PARAM_HOOK_FN,
     "deepseek3.2-671b": DEEPSEEK_MAXTEXT_TO_HF_PARAM_HOOK_FN,
-    "deepseek4-tiny": DEEPSEEKV4_MAXTEXT_TO_HF_PARAM_HOOK_FN,
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/utils/param_mapping.py:4351]: `src/maxtext/configs/types.py` and `tests/unit/deepseek_v4_vs_reference_test.py` still contains reference to deepseek4-tiny

File: src/maxtext/checkpoint_conversion/utils/utils.py [hunk 6]
```diff
@@ -1285,8 +1285,9 @@ def save_weights_to_checkpoint(
   if checkpoint_manager is None:
     raise RuntimeError("Failed to create Orbax checkpoint manager.")
 
+  params = jax_weights if (isinstance(jax_weights, dict) and "params" in jax_weights) else {"params": jax_weights}
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/utils/utils.py:1299]: While preserving top-level collections (`Tid2EidVar`, `MoEBiasVar`) in `state_new.params` works for `to_maxtext.py`, it breaks `to_huggingface.py`.

In `to_huggingface.py:529`, checkpoint extraction calls `detect_and_extract_checkpoint()` in `utils.py:1093`:
```python
if isinstance(actual_weights_dict, dict) and "params" in actual_weights_dict:
  actual_weights_dict = actual_weights_dict["params"]
return extract_linen_weights(actual_weights_dict)

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py (hunk 5 above)
Comment (snehalv2002) [reply src/maxtext/checkpoint_conversion/utils/param_mapping.py:4351]: Restored `deepseek4-tiny` in `HOOK_FNS` and added it to `PARAM_MAPPING` and `HF_WEIGHTS_TO_SHAPE` so `deepseek4-tiny` remains fully registered and supported for unit testing, consistent with `types.py` and `tests/unit/deepseek_v4_vs_reference_test.py`.

**Review by AusarYao28 — COMMENTED** [review]
Please add unit test cases for `DEEPSEEK_V4_MAXTEXT_TO_HF_PARAM_MAPPING` and `DEEPSEEK_V4_MAXTEXT_TO_HF_PARAM_HOOK_FN` (covering both scanned and unscanned modes) to `tests/unit/param_mapping_test.py` so that DeepSeek-V4 roundtrip mapping is covered by CI.

File: src/maxtext/checkpoint_conversion/utils/param_mapping.py [hunk 7]
```diff
@@ -3927,328 +3927,282 @@ def reshape_vision_attn_out(input_tensor, target_shape):
...
+    return get_hf_key
 
   if not scan_layers:
     for i in range(n_layers):
-      add_layer_mapping(f"params-decoder-layers_{i}", i)
+      _add_layer(
+          f"params-decoder-layers_{i}",
+          _make_unrolled_hf_key_fn(i),
+          is_hash=(i < num_hash_layers),
+          compressor_type="both",
```
Comment (AusarYao28) [inline src/maxtext/checkpoint_conversion/utils/param_mapping.py]: Setting `compressor_type="both"` for prefix layers and all unscanned layers generates phantom mapping keys:
- Layers 0 and 1 have no compressor (`compress_ratio = 0`).
- Layer 2 has CSA only (`compress_ratio = 4`).
- Subsequent layers alternate between CSA and HCA. No layer ever has both.
While `validate_and_filter_param_map_keys` currently swallows these extra keys, we should avoid injecting nonexistent keys into `PARAM_MAPPING`. Please use `compress_ratios` dynamically here similar to `hf_shape.py:245`.

**Review by AusarYao28 — APPROVED** [review]

**Review by RissyRan — APPROVED** [review]
LGTM, could you post your test results in the PR description?
